### PR TITLE
feat(x/ecocredit): params ORM

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

setup for params to use ORM and new msg based gov params

not sure if this should be wired in yet? pending .046 SDK 

Closes: #729 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1/tx.proto [hunk 1]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+  repeated Fee remove_fees = 3;
+
+  // Fee represents a fee to be used in the credit class creation fee.
+  message Fee {
+    // denom is the coin denom
+    string denom = 1;
+    // amount is the amount of the coin needed to create the credit class.
+    string amount = 2;
+  }
+}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:393]: This is a new feature that we might need to further discuss on a product call. I think for now we want to have REGEN be the only token accepted for fees on our network.

Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:393]: This is a new feature that we might need to further discuss on a product call. I think for now we want to have REGEN be the only token accepted for fees on our network.

**Review by ryanchristo — CHANGES_REQUESTED** [review]
Looking good although I don't think we should merge this until we are certain v0.46 will land in the next release and therefore I'm going to block this for now. We should at least wait until a beta release for Hambach.

We should probably discuss the fee lists on a product call and, if we want to support fee lists, whether they should be the same list. I'm not sure we do though.

Also, we should remove `proto/regen/ecocredit/v1/types.proto` as this only includes the `Params` proto message. Any reason for keeping it in this pull request?

File: proto/regen/ecocredit/v1/tx.proto [hunk 2]
```diff
@@ -57,6 +73,11 @@ message MsgCreateClass {
 
   // credit_type_abbrev describes the abbreviation of a credit type (e.g. "C", "BIO").
   string credit_type_abbrev = 4;
+
+  // fee_denom is the currency denom to be used to pay the credit class fee.
+  // This denom MUST exist in the governance controlled allowed credit class fee list.
+  // The user's balance must be at least equal to the amount specified in the credit class fee list.
+  string fee_denom = 5;
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply proto/regen/ecocredit/v1/tx.proto:80]: the existing param for credit class fee is of type `Coins`, which is a list of coins 🤷🏻

File: x/ecocredit/server/core/keeper_test.go [hunk 3]
```diff
@@ -35,6 +35,7 @@ type baseSuite struct {
 	addr         sdk.AccAddress
 	bankKeeper   *mocks.MockBankKeeper
 	paramsKeeper *mocks2.MockParamKeeper
+	accountKeeper 	*mocks2.MockAccountKeeper
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply x/ecocredit/server/core/keeper_test.go:38]: hmm yeah this keeps happening to my IDE - anytime i make a change it says i need to `go mod tidy` and then doesnt run go fmt... does this happen to you?

File: proto/regen/ecocredit/v1/tx.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanchristo) [reply proto/regen/ecocredit/v1/tx.proto:80]: Ah, you're right. Maybe this is ok then. For some reason I thought it was `Coin`.

File: proto/regen/ecocredit/v1/tx.proto (hunk 1 above)
Comment (ryanchristo) [reply proto/regen/ecocredit/v1/tx.proto:393]: Overlooked how the param is currently set up.

File: proto/regen/ecocredit/v1/tx.proto [hunk 4]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+  repeated Fee remove_fees = 3;
+
+  // Fee represents a fee to be used in the credit class creation fee.
+  message Fee {
+    // denom is the coin denom
+    string denom = 1;
+    // amount is the amount of the coin needed to create the credit class.
+    string amount = 2;
+  }
+}
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply proto/regen/ecocredit/v1/tx.proto:417]: Same here. Overlooked how the param is currently set up.

File: x/ecocredit/server/core/keeper_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/ecocredit/server/core/keeper_test.go]: Hmmm... whitespace is fixed for me on save. Also, I'm running `make tidy` (`./scripts/go-mod-tidy-all.sh` under the hood).

File: x/ecocredit/server/core/params.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+import (
+	"context"
+	"github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	ecocreditv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+func (k Keeper) NewCreditType(ctx context.Context, req *core.MsgNewCreditTypeRequest) (*core.MsgNewCreditTypeResponse, error) {
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply x/ecocredit/server/core/params.go]: changed https://github.com/regen-network/regen-ledger/pull/854/commits/a9d202f1533fe31f629575cfe2dc6185467a15e6

**Comment by technicallyty** [comment]
> Also, we should remove `proto/regen/ecocredit/v1/types.proto` as this only includes the `Params` proto message. Any reason for keeping it in this pull request?


i was unsure if this was going to make it in the next release, so i left existing params in place. mostly wanted to introduce the new params API to see how folks like this. i think theres other ways to update params (one monotlithic update function with onofs, etc).

File: proto/regen/ecocredit/v1/state.proto [hunk 6]
```diff
@@ -220,3 +221,57 @@ message BatchSupply {
...
+
+    // address is the address of the user that is allowed to create credit classes
+    bytes address = 1;
+}
+
+// AllowlistEnabled is a governance controlled, singleton table that holds a boolean value.
+// When set to true, only the addresses in AllowedClassCreators can create credit classes.
+// When set to false, any address is free to create credit classes.
+message AllowlistEnabled {
+    option(cosmos.orm.v1alpha1.singleton) = {
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/state.proto:241]:
```suggestion
// AllowlistEnabled determines if the credit class creator allowlist is enabled.
// When set to true, only the addresses in the AllowedClassCreator table can create
// credit classes. When set to false, any address is free to create credit classes.
message AllowlistEnabled {
    // This table is controlled via governance.
    option (cosmos.orm.v1alpha1.singleton) = {
```

File: proto/regen/ecocredit/v1/state.proto [hunk 7]
```diff
@@ -220,3 +221,57 @@ message BatchSupply {
...
+    bool enabled = 1;
+}
+
+// CreditClassFee stores the valid coin denoms and amounts that may be used as the fee for credit class creation.
+// This is controlled via governance.
+message CreditClassFee {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 13,
+        primary_key: {fields: "denom"}
+    };
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/state.proto:255]:
```suggestion
// CreditClassFee is a valid coin denom and amount that may be used as the fee for
// credit class creation.
message CreditClassFee {
    // This table is controlled via governance.
    option (cosmos.orm.v1alpha1.table) = {
        id: 13,
        primary_key: {fields: "denom"}
    };
```

File: proto/regen/ecocredit/v1/state.proto [hunk 8]
```diff
@@ -220,3 +221,57 @@ message BatchSupply {
...
+    string amount = 2;
+}
+
+// BasketFee stores the valid coin denoms and amounts that may be used as the fee for creating a basket.
+// This is controlled via governance.
+message BasketFee {
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 14,
+        primary_key: {fields: "denom"}
+    };
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/state.proto:270]:
```suggestion
// BasketFee is a valid coin denom and amount that may be used as the fee for
// creating a basket.
message BasketFee {
    // This table is controlled via governance.
    option (cosmos.orm.v1alpha1.table) = {
        id: 14,
        primary_key: {fields: "denom"}
    };
```

File: proto/regen/ecocredit/v1/state.proto [hunk 9]
```diff
@@ -220,3 +221,57 @@ message BatchSupply {
...
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 13,
+        primary_key: {fields: "denom"}
+    };
+
+    // denom is the denom of the coin allowed to be used for credit class fees
+    string denom = 1;
+
+    // amount is the fee amount
+    string amount = 2;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/state.proto:261]:
```suggestion
    // denom is the denom of the fee required to create a credit class.
    string denom = 1;

    // amount is the amount of the fee required to create a credit class.
    string amount = 2;
```

File: proto/regen/ecocredit/v1/state.proto [hunk 10]
```diff
@@ -220,3 +221,57 @@ message BatchSupply {
...
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 14,
+        primary_key: {fields: "denom"}
+    };
+
+    // denom is the denom of the coin allowed to be used for basket creation fees
+    string denom = 1;
+
+    // amount is the fee amount
+    string amount = 2;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/state.proto:276]:
```suggestion
    // denom is the denom of the fee required to create a basket.
    string denom = 1;

    // amount is the amount of the fee required to create a basket.
    string amount = 2;
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 11]
```diff
@@ -42,6 +43,21 @@ service Msg {
...
+  rpc ToggleAllowList(MsgToggleAllowListRequest) returns (MsgToggleAllowListResponse);
+
+  // UpdateAllowedCreditClassCreators updates the allow list of credit class creators - Governance only method
+  rpc UpdateAllowedCreditClassCreators(MsgUpdateAllowedCreditClassCreatorsRequest) returns (MsgUpdateAllowedCreditClassCreatorsResponse);
+
+  // UpdateCreditClassFee updates the list of allowed denoms and their amounts to be used as credit class fees - Governance only method
+  rpc UpdateCreditClassFee(MsgUpdateCreditClassFeeRequest) returns (MsgUpdateCreditClassFeeResponse);
+
+  // UpdateBasketFee updates the list of allowed denoms and their amounts to be used as basket creation fees - Governance only method
+  rpc UpdateBasketFee(MsgUpdateBasketFeeRequest) returns (MsgUpdateBasketFeeResponse);
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:60]:
```suggestion
  // AddCreditType adds a new credit type. This message must be signed by the governance module.
  rpc AddCreditType(MsgAddCreditType) returns (MsgAddCreditTypeResponse);

  // ToggleAllowList toggles the allow list. This message must be signed by the governance module.
  rpc ToggleAllowList(MsgToggleAllowListRequest) returns (MsgToggleAllowListResponse);

  // UpdateAllowedCreditClassCreators updates the allow list of credit class creators. This message
  // must be signed by the governance module.
  rpc UpdateAllowedCreditClassCreators(MsgUpdateAllowedCreditClassCreatorsRequest) returns (MsgUpdateAllowedCreditClassCreatorsResponse);

  // UpdateCreditClassFee updates the list of allowed denoms and their amounts to be used as credit
  // class fees. This message must be signed by the governance module.
  rpc UpdateCreditClassFee(MsgUpdateCreditClassFeeRequest) returns (MsgUpdateCreditClassFeeResponse);

  // UpdateBasketFee updates the list of allowed denoms and their amounts to be used as basket
  // creation fees. This message must be signed by the governance module.
  rpc UpdateBasketFee(MsgUpdateBasketFeeRequest) returns (MsgUpdateBasketFeeResponse);
```

File: proto/regen/ecocredit/v1/tx.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanchristo) [reply proto/regen/ecocredit/v1/tx.proto:80]: Looks like this will need to be resolved with master now.

File: proto/regen/ecocredit/v1/tx.proto [hunk 12]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+// MsgAddCreditType is the Msg/AddCreditType request type.
+// The credit types are immutable and cannot be deleted.
+message MsgAddCreditType {
+  // credit_types is a list of new credit types to add
+  repeated CreditType credit_types = 1;
+
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
+  string root_address = 2;
+}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:339]: Given this is the message for add, I don't think we need to include "The credit types are immutable and cannot be deleted." Also this should either be plural (we are adding multiple credit types) or we should limit to one credit type at a time.

```suggestion
// MsgAddCreditTypes is the Msg/AddCreditTypes request type.
message MsgAddCreditTypes {
  // credit_types is a list of credit types to add.
  repeated CreditType credit_types = 1;

  // root_address is the address of the signer.
  // This MUST equal the address of the gov module for the tx to succeed.
  string root_address = 2;
}
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 13]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+
+// MsgToggleAllowListRequest is the Msg/ToggleAllowListRequest request type
+message MsgToggleAllowListRequest {
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
+  string root_address = 1;
+
+  // toggle is what the allow list will be set to.
+  bool toggle = 2;
+}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:353]:
```suggestion
// MsgToggleAllowListRequest is the Msg/ToggleAllowListRequest request type
message MsgToggleAllowListRequest {
  // root_address is the address of the signer.
  // This MUST equal the address of the gov module for the tx to succeed.
  string root_address = 1;

  // toggle is what the allow list will be set to.
  bool toggle = 2;
}
```

Comment (ryanchristo) [reply proto/regen/ecocredit/v1/tx.proto:353]: Might also be nice to consistently put root address first or last in these message.

File: proto/regen/ecocredit/v1/tx.proto [hunk 14]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+
+// MsgToggleAllowListResponse is the Msg/MsgToggleAllowListResponse response type.
+message MsgToggleAllowListResponse {}
+
+// MsgUpdateAllowedCreditClassCreatorsRequest is the Msg/UpdateAllowedCreditClassCreators request type
+message MsgUpdateAllowedCreditClassCreatorsRequest {
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:361]:
```suggestion
  // root_address is the address of the signer.
  // This MUST equal the address of the gov module for the tx to succeed.
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 15]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+}
+
+// MsgNewCreditClassCreatorResponse is the Msg/UpdateAllowedCreditClassCreators response type
+message MsgUpdateAllowedCreditClassCreatorsResponse {}
+
+// MsgUpdateCreditClassFeeRequest is the Msg/UpdateCreditClassFee request type
+message MsgUpdateCreditClassFeeRequest {
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
+  string root_address = 1;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:378]:
```suggestion
// MsgUpdateCreditClassFeesRequest is the Msg/UpdateCreditClassFees request type
message MsgUpdateCreditClassFeesRequest {
  // root_address is the address of the signer.
  // This MUST equal the address of the gov module for the tx to succeed.
  string root_address = 1;
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 16]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+  // credit_types is a list of new credit types to add
+  repeated CreditType credit_types = 1;
+
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
+  string root_address = 2;
+}
+
+// MsgAddCreditTypeResponse is the Msg/AddCreditType response type.
+message MsgAddCreditTypeResponse {}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:342]:
```suggestion
// MsgAddCreditTypesResponse is the Msg/AddCreditTypes response type.
message MsgAddCreditTypesResponse {}
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 17]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+  message Fee {
+    // denom is the coin denom
+    string denom = 1;
+    // amount is the amount of the coin needed to create the credit class.
+    string amount = 2;
+  }
+}
+
+// MsgUpdateCreditClassFeeResponse is the Msg/UpdateCreditClassFee response type
+message MsgUpdateCreditClassFeeResponse {}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:396]:
```suggestion
// MsgUpdateCreditClassFeesResponse is the Msg/UpdateCreditClassFees response type
message MsgUpdateCreditClassFeesResponse {}
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 18]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+
+// MsgUpdateBasketFeeRequest is the Msg/UpdateBasketFee request type.
+message MsgUpdateBasketFeeRequest {
+  // root_address is the address of the caller.
+  // this MUST equal the address of the gov module for the tx to succeed.
+  string root_address = 1;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:402]:
```suggestion
// MsgUpdateBasketFeesRequest is the Msg/UpdateBasketFees request type.
message MsgUpdateBasketFeesRequest {
  // root_address is the address of the caller.
  // This MUST equal the address of the gov module for the tx to succeed.
  string root_address = 1;
```

File: proto/regen/ecocredit/v1/tx.proto [hunk 19]
```diff
@@ -305,3 +326,95 @@ message MsgUpdateClassMetadata {
...
+  message Fee {
+    // denom is the coin denom
+    string denom = 1;
+    // amount is the amount of the coin needed to create the credit class.
+    string amount = 2;
+  }
+}
+
+// MsgUpdateBasketFeeResponse is the Msg/UpdateBasketFee response type.
+message MsgUpdateBasketFeeResponse {}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:420]:
```suggestion
// MsgUpdateBasketFeesResponse is the Msg/UpdateBasketFees response type.
message MsgUpdateBasketFeesResponse {}
```

**Review by ryanchristo — COMMENTED** [review]
Unblocking this as it does not remove existing param functionality. I only reviewed the proto files to start. Looking ahead a bit, I'm not sure if the implementation should exist within the "core" directory but instead at the root level since the params are not specific to "core", i.e. we have a basket fees param.

**Pull request closed without merging**